### PR TITLE
[PD-2086]  History support

### PR DESCRIPTION
**Pull request by darrint**
Replace the ad-hoc routing with no history support in DynamicReportApp with react-router.

This PR may be easier to review commit by commit. The commit "Refactor reporting wizard" touches a lot of files. The other commits are small.

The refactor does a couple of things:

* Make each page responsible for loading its own data.
* Adds the Router.
* ReportEngine's interface for notifying a react component of newly created reports moved to ReportFinder. This simplified routing.

### CODE REVIEW

File: gulp/src/common/ui/Loading.jsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/* global spinnerImg */
```
Comment (lemniscate613) [inline gulp/src/common/ui/Loading.jsx]: Nice. I'd like to use this component.

**Comment by lemniscate613** [comment]
Not sure where this falls in scope:

The tag system is definitely complex behind the scenes. It took me awhile to realize, though, you aren't supposed to click or drag the tags in the top section. I think its only for exploring the total possible tags (right?). If you find a tag you like, you should clear the search, click the green plus button, redo your search, and add the tag there. This doesn't behave like any tag system I've ever worked with.

screen shot 2016-01-26 at 9 50 48 am

**Comment by lemniscate613** [comment]
Not sure where this falls in scope:

I entered the date: 01/01/1990 and the server complained:

```
ERROR:django.request:Internal Server Error: /reports/api/help
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/MyJobs/myjobs/decorators.py", line 241, in wrap
    return view_func(request, *args, **kwargs)
  File "/MyJobs/universal/decorators.py", line 109, in wrap
    return view_func(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/http.py", line 41, in inner
    return func(request, *args, **kwargs)
  File "/MyJobs/import_jobs/../myreports/views.py", line 525, in help_api
    result = driver.help(company, filter_spec, field, partial)
  File "/MyJobs/import_jobs/../myreports/datasources/jsondriver.py", line 31, in help
    filter_obj = self.build_filter(filter_spec)
  File "/MyJobs/import_jobs/../myreports/datasources/jsondriver.py", line 53, in build_filter
    for d in date_strings]
  File "/usr/lib/python2.7/_strptime.py", line 325, in _strptime
    (data_string, format))
ValueError: time data '01/01/1990' does not match format '%Y-%m-%d'
```

**Comment by tdphillips** [comment]
https://directemployersfoundation.atlassian.net/browse/PD-2062

**Comment by darrint** [comment]
Regarding the tags interface, @misterjonscott is redesigning that interface on this ticket: https://directemployersfoundation.atlassian.net/browse/PD-2106

**Pull request merged**